apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
-   * is equal to the value for the current panel's arguments' same property.
-   */
-  isCurrentPanel: function n_isCurrentPanel(panel, args) {
-    if (!currentPanel || !panel) {
-      return false;
-    }
+  return defer.promise.then(catchStepError);
+}
 
-    if (panel !== currentPanel.panel) {
-      return false;
-    }
+/**
+ * Used in `waitForSlideAnimation` to help solving race conditions.
+ * @type {Defer}
+ */
+var previousAnimationDefer;
+
+/**
+ * Wait for the end of an animation on `panelElement`.
+ *
+ * @param {Element} panelElement The element to watch.
+ * @returns {Promise} Resolves when the animation ends.
+ */
+function waitForSlideAnimation(panelElement) {
+  if (previousAnimationDefer) {
+    previousAnimationDefer.reject(new Error('A new animation started'));
+  }
 
-    if (!args) {
-      return true;
-    }
+  var defer = Utils.Promise.defer();
+  previousAnimationDefer = defer;
 
-    var currentArgs = currentPanel.args || {};
+  panelElement.addEventListener('animationend', function onAnimationEnd() {
+    this.removeEventListener('animationend', onAnimationEnd);
+    defer.resolve();
+  });
 
-    for (var arg in args) {
-      if (args[arg] != currentArgs[arg]) {
-        return false;
-      }
-    }
+  return defer.promise.then(() => previousAnimationDefer = null);
+}
 
-    return true;
-  },
+/**
+ * Hides current panel, shows new panel.
+ * It will use an animation if we change panels in the same document.
+ *
+ * @returns {Promise} Resolves when the animation is over, if there is an
+ * animation. Otherwise returns a resolved promise.
+ */
+function switchPanel() {
+  if (!navigationTransition) {
+    throw new Error(
+      'Trying to switch panel but no transition is started yet.'
+    );
+  }
 
-  /**
-   * Checks if current location.hash corresponds to default panel.
-   * @returns {boolean} True if current location.hash corresponds to default
-   * panel.
-   */
-  isDefaultPanel: function n_isDefaultPanel() {
-    var panelName = this.getPanelName();
+  var from = navigationTransition.states.from;
+  var to = navigationTransition.states.to;
 
-    return !panelName || panelName === this.defaultPanel;
-  },
+  var oldView, newView;
+  var oldPanelElement, newPanelElement;
 
-  /**
-   * Ensures that current panel is correctly set, if no current panel set and
-   * we're not in the process of transitioning then set default panel as current
-   * one.
-   */
-  ensureCurrentPanel: function() {
-    if (this.transitionPromise) {
-      return this.transitionPromise;
-    }
+  if (from) {
+    oldView = from.panel;
+    oldPanelElement = document.querySelector(`.panel-${oldView.name}`);
+  }
 
-    if (!currentPanel) {
-      return this.toPanel(this.defaultPanel);
-    }
+  newView = to.panel;
+  newPanelElement = document.querySelector(`.panel-${newView.name}`);
 
-    return Promise.resolve();
+  var doSlideAnimation = oldView && newView &&
```

Comment:
Mmm, it looks like ```newView``` should always be defined otherwise template string above (```newPanelElement = document.querySelector(`.panel-${newView.name}`)```) will throw. Is there any case when ```newView``` can be null/undefined?